system/jhtree/jhtree.cpp

```diff
@@ -2614,14 +2613,26 @@ const CJHTreeNode *CNodeCache::getNode(const INodeLoader *keyIndex, unsigned iD,
     CNodeMRUCache & curCache = cache[cacheType];
     CriticalSection & cacheLock = lock[cacheType];
     Owned<CNodeCacheEntry> ownedCacheEntry; // ensure node gets cleaned up if it fails to load
+    const CJHTreeNode * fastPathMatch;
     bool alreadyExists = true;
     {
         CNodeCacheEntry * cacheEntry;
         unsigned hashcode = curCache.getKeyHash(key);
-        CriticalBlock block(cacheLock);
 
+        CriticalBlock block(cacheLock);
         cacheEntry = curCache.query(hashcode, &key);
-        if (unlikely(!cacheEntry))
+        if (likely(cacheEntry))
+        {
+            fastPathMatch = cacheEntry->queryNode();
+            if (likely(fastPathMatch))
+            {
+                //Avoid linking and releasing  cacheEntry if the match is already loaded.  The goto is ugly, but
+                //anything that reduces the time the critical section is held will help significantly
+                fastPathMatch->Link();
+                goto returnFastPath;  // the goto is horrible, but make sure atomics are incremented outside the critical section
```

Comment:
An alternative to using a goto would be to use a LeavableCriticalBlock.  I suspect a good compiler would generate the same code for both.